src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -257,36 +316,46 @@ public boolean isFingerprintArtifacts() {
     }
 
     @Override
-    public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener)
-            throws InterruptedException, IOException {
-        upgradeIfNecessary(build.getProject());
+    public void perform(@Nonnull Run<?, ?> build, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        if (build instanceof AbstractBuild) {
+            upgradeIfNecessary(((AbstractBuild)build).getProject());
+        }
+
+        EnvVars env = build.getEnvironment(listener);
+        if (build instanceof AbstractBuild) {
+            env.overrideAll(((AbstractBuild)build).getBuildVariables()); // Add in matrix axes..
+        }
+
         PrintStream console = listener.getLogger();
         String expandedProject = project, expandedFilter = filter;
         String expandedExcludes = getExcludes();
         try {
-            EnvVars env = build.getEnvironment(listener);
-            env.overrideAll(build.getBuildVariables()); // Add in matrix axes..
             expandedProject = env.expand(project);
-            Job<?, ?> job = Jenkins.getInstance().getItem(expandedProject, build.getProject().getRootProject().getParent(), Job.class);
+            Job<?, ?> job = Jenkins.getInstance().getItem(expandedProject, getItemGroup(build), Job.class);
             if (job != null && !expandedProject.equals(project)
                 // If projectName is parameterized, need to do permission check on source project.
                 && !canReadFrom(job, build)) {
                 job = null; // Disallow access
             }
             if (job == null) {
-                console.println(Messages.CopyArtifact_MissingProject(expandedProject));
-                return false;
+                String message = Messages.CopyArtifact_MissingProject(expandedProject);
+                console.println(message);
+                throw new CopyArtifactException(message);
```

Comment:
If you are throwing an `AbortException`, it is redundant to print the message. Just

```
if (job == null) {
    throw new AbortException(Messages.CopyArtifact_MissingProject(expandedProject));
}
```